File: components/goodbye/src/test/java/io/undertree/goodbye/GoodbyeControllerTests.java
package io.undertree.goodbye;

import java.util.Optional;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class GoodbyeControllerTests {

    @Mock
    GoodbyeRepository goodbyeRepository;

    @Test
    public void given_GoodbyeController_WhenGoodbyeRepositoryFindByIdIs1_Then_ExpectOkRalph() throws Exception {
        when(goodbyeRepository.findById(1L))
                .thenReturn(Optional.of(new GoodbyeRecord(1L, "Ralph")));

        GoodbyeController controller = new GoodbyeController(goodbyeRepository);
        ResponseEntity<GoodbyeRecord> response = controller.goodbye(1L);

        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
        assertThat(response.getBody()).isNotNull();
        assertThat(response.getBody().getId()).isEqualTo(1L);
        assertThat(response.getBody().getName()).isEqualTo("Ralph");
    }
}

File: components/goodbye/src/test/java/io/undertree/goodbye/GoodbyeControllerTests.kt
package io.undertree.goodbye

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.BDDMockito.anyLong
import org.mockito.BDDMockito.given
import org.mockito.Mock
import org.mockito.runners.MockitoJUnitRunner
import org.springframework.http.HttpStatus
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class GoodbyeControllerTests {

    @Mock
    lateinit var goodbyeRepository: GoodbyeRepository

    @Test
    fun `Should expect 200 and body containing id and name`() {

        given(goodbyeRepository.findById(1L))
                .willReturn(Optional.of(GoodbyeEntity(1L, "Ralph")))

        val controller = GoodbyeController(goodbyeRepository)
        val response = controller.oneById(1L)

        assertThat(response.statusCode).isEqualTo(HttpStatus.OK)
        assertThat(response.body).isNotNull()
        assertThat(response.body.id).isEqualTo(1L)
        assertThat(response.body.name).isEqualTo("Ralph")
    }

    @Test
    fun `Should expect 404`() {

        given(goodbyeRepository.findById(anyLong()))
                .willReturn(Optional.empty())

        val controller = GoodbyeController(goodbyeRepository)
        val response = controller.oneById(1L)

        assertThat(response.statusCode).isEqualTo(HttpStatus.NOT_FOUND)
    }
}
